### PR TITLE
porting diag_second_shoc_moments

### PR DESCRIPTION
**Pull request by xyuan**
porting diag_second_shoc_moments

[BFB]

### CODE REVIEW

**Review by jeff-cohere — APPROVED** [review]

File: components/scream/src/physics/shoc/shoc_diag_second_shoc_moments_impl.hpp [hunk 1]
```diff
@@ -13,18 +13,74 @@ namespace shoc {
...
 
+  // Calculate surface properties needed for lower
+  //  boundary conditions
+  shoc_diag_second_moments_srf(wthl_sfc, uw_sfc, vw_sfc, ustar2, wstar);
+  team.team_barrier();
 
+  // Diagnose the second order moments flux, for the lower boundary
+  const auto nlevi_pack = ekat::npack<Spack>(nlevi)-1;
+  const int nlevi_indx  = (nlevi-1)%Spack::n;
+  Scalar wthl_s         = wthl_sec(nlevi_pack)[nlevi_indx];
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_diag_second_shoc_moments_impl.hpp:37]: This code is significantly more complicated than it needs to be. If you just pass these outputs directly (they are taken by reference), then you don't need all these temporaries.

```
shoc_diag_second_moments_lbycond(wthl_sfc, wqw_sfc, uw_sfc, vw_sfc, ustar2, wstar,
          wthl_sec(nlevi_pack)[nlevi_indx], 
          wqw_sec(nlevi_pack)[nlevi_indx],
          uw_sec(nlevi_pack)[nlevi_indx],
          vw_sec(nlevi_pack)[nlevi_indx],
          wtke_sec(nlevi_pack)[nlevi_indx],
          thl_sec(nlevi_pack)[nlevi_indx],
          qw_sec(nlevi_pack)[nlevi_indx],
          qwthl_sec(nlevi_pack)[nlevi_indx]);
```

File: components/scream/src/physics/shoc/shoc_diag_second_shoc_moments_impl.hpp [hunk 2]
```diff
@@ -13,18 +13,74 @@ namespace shoc {
...
+  // Diagnose the second order moments, calculate the upper boundary conditions
+  Scalar thl_s0   = thl_sec(0)[0];
+  Scalar qw_s0    = qw_sec(0)[0];
+  Scalar wthl_s0  = wthl_sec(0)[0];
+  Scalar wqw_s0   = wqw_sec(0)[0];
+  Scalar qwthl_s0 = qwthl_sec(0)[0];
+  Scalar uw_s0    = uw_sec(0)[0];
+  Scalar vw_s0    = vw_sec(0)[0];
+  Scalar wtke_s0  = wtke_sec(0)[0];
+  shoc_diag_second_moments_ubycond(thl_s0, qw_s0, wthl_s0, wqw_s0, qwthl_s0, uw_s0, vw_s0, wtke_s0);
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_diag_second_shoc_moments_impl.hpp:74]: Same comment as above.

File: components/scream/src/physics/shoc/tests/shoc_diag_second_shoc_moments_tests.cpp [hunk 3]
```diff
@@ -292,7 +298,9 @@ struct UnitWrap::UnitTest<D>::TestDiagSecondShocMoments {
     // Get data from cxx
     for (auto& d : cxx_data) {
       d.transpose<ekat::TransposeDirection::c2f>(); // _f expects data in fortran layout
-      diag_second_shoc_moments_f(d.shcol(), d.nlev(), d.nlevi(), d.thetal, d.qw, d.u_wind, d.v_wind, d.tke, d.isotropy, d.tkh, d.tk, d.dz_zi, d.zt_grid, d.zi_grid, d.shoc_mix, d.wthl_sfc, d.wqw_sfc, d.uw_sfc, d.vw_sfc, d.thl_sec, d.qw_sec, d.wthl_sec, d.wqw_sec, d.qwthl_sec, d.uw_sec, d.vw_sec, d.wtke_sec, d.w_sec);
+      diag_second_shoc_moments_f(d.shcol(), d.nlev(), d.nlevi(), d.thetal, d.qw, d.u_wind, d.v_wind, d.tke, d.isotropy,
```
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_diag_second_shoc_moments_tests.cpp:301]: Please upstream merge this branch. After that, shcol, nlev, and nlevi are no longer methods, just access them directly.

**Review by jgfouca — CHANGES_REQUESTED** [review]
Needs upstream merge and some changes to the core impl

File: components/scream/src/physics/shoc/shoc_diag_second_shoc_moments_impl.hpp (hunk 2 above)
Comment (xyuan) [reply components/scream/src/physics/shoc/shoc_diag_second_shoc_moments_impl.hpp:66]: Yes, done!

**Comment by xyuan** [comment]
@jgfouca @tcclevenger This PR is good to go, please review it.

File: components/scream/src/physics/shoc/shoc_diag_second_shoc_moments_impl.hpp [hunk 4]
```diff
@@ -13,18 +13,57 @@ namespace shoc {
...
+  // Diagnose the second order moments, calculate the upper boundary conditions
+  Scalar& thl_s0   = thl_sec(0)[0];
+  Scalar& qw_s0    = qw_sec(0)[0];
+  Scalar& wthl_s0  = wthl_sec(0)[0];
+  Scalar& wqw_s0   = wqw_sec(0)[0];
+  Scalar& qwthl_s0 = qwthl_sec(0)[0];
+  Scalar& uw_s0    = uw_sec(0)[0];
+  Scalar& vw_s0    = vw_sec(0)[0];
+  Scalar& wtke_s0  = wtke_sec(0)[0];
+  shoc_diag_second_moments_ubycond(thl_s0, qw_s0, wthl_s0, wqw_s0, qwthl_s0, uw_s0, vw_s0, wtke_s0);
```
Comment (tcclevenger) [inline components/scream/src/physics/shoc/shoc_diag_second_shoc_moments_impl.hpp:66]: I think Jim is saying to not create temporary variables, but instead do something like this:
```suggestion
  shoc_diag_second_moments_ubycond(thl_sec(0)[0], qw_sec(0)[0], wthl_sec(0)[0], wqw_sec(0)[0], qwthl_sec(0)[0], uw_sec(0)[0], vw_sec(0)[0], wtke_sec(0)[0]);
```

Comment (xyuan) [reply components/scream/src/physics/shoc/shoc_diag_second_shoc_moments_impl.hpp]: OK, sounds good to me.

**Review by tcclevenger — APPROVED** [review]

File: components/scream/scripts/gather_all_data.py [hunk 5]
```diff
@@ -3,7 +3,7 @@
 import os, pathlib
 import concurrent.futures as threading3
 from machines_specs import get_mach_env_setup_command, get_mach_batch_command, get_mach_testing_resources, \
-                           get_mach_cxx_compiler, get_mach_f90_compiler, get_mach_c_compiler
+                           get_mach_cxx_compiler, get_mach_f90_compiler, get_mach_c_compiler, is_cuda_machine
```
Comment (jgfouca) [inline components/scream/scripts/gather_all_data.py:6]: Wondering why github is attributing these changes to your PR? Did something go wrong with the upstream merge?

**Comment by jgfouca** [comment]
@xyuan , yep, something's wrong with the history. You'll need to remake this branch.

**Comment by jgfouca** [comment]
My guess is that you rebased new master commits into your branch

**Comment by xyuan** [comment]
Yes, I did rebase master and merge it to this branch. Maybe I need to redo this PR.

**Comment by xyuan** [comment]
It shows conflict with each other when I did the rebase, so I manually edit it.

**Comment by jgfouca** [comment]
@xyuan , yeah, you cannot mix upstream merges and rebases.

**Comment by xyuan** [comment]
Close this PR, since I am going to redo it.

**Pull request closed without merging**